### PR TITLE
Features 201, 202 and 203

### PR DESCRIPTION
**Pull request by jicarretero**
- Feature 201 - IoT Sensors implemented and changed environment.py
- Feature 201 - IoT Sensors implemented and changed environment.py
- Partially feature 202 included
- Up to point 19 of Tutorial 202
- Feature 202 finished
- Feature 203 finished

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -117,3 +117,8 @@ xunit.xml
 
 # Plain test results
 features/*/plain.output
+
```
Comment (flopezag) [inline .gitignore:120]: why those content in the .gitignore?

File: features/data/202.IoT_Agent_Ultralight/tags [hunk 2]
```diff
@@ -0,0 +1,367 @@
+!_TAG_FILE_SORTED	2	/0=unsorted, 1=sorted, 2=foldcase/
```
Comment (flopezag) [inline features/data/202.IoT_Agent_Ultralight/tags]: Do we need it?

File: features/data/203.IoT_Agent_Json/tags [hunk 3]
```diff
@@ -0,0 +1,137 @@
+!_TAG_FILE_SORTED	2	/0=unsorted, 1=sorted, 2=foldcase/
```
Comment (flopezag) [inline features/data/203.IoT_Agent_Json/tags]: do we need it?

File: features/environment.py [hunk 4]
```diff
@@ -24,16 +57,29 @@ def before_feature(context, feature):
     stdout.write("=========== START FEATURE ===========\n")
     stdout.write(f'Feature name: {feature.name}\n\n')
 
-    parameters = [s for s in feature.description if 'docker-compose' in s or 'environment' in s]
-    parameters = dict(s.split(':', 1) for s in parameters)
+    p = [s for s in feature.description if is_interesting_feature_string(s)]
```
Comment (flopezag) [inline features/environment.py:60]: why not simply:

[s for s in feature.description if s in INTERESTING_FEATURES_STRINGS]

File: features/environment.py [hunk 5]
```diff
@@ -24,16 +57,29 @@ def before_feature(context, feature):
     stdout.write("=========== START FEATURE ===========\n")
     stdout.write(f'Feature name: {feature.name}\n\n')
 
-    parameters = [s for s in feature.description if 'docker-compose' in s or 'environment' in s]
-    parameters = dict(s.split(':', 1) for s in parameters)
+    p = [s for s in feature.description if is_interesting_feature_string(s)]
+    parameters = {}
+    # parameters = dict(s.split(':', 1) for s in parameters)
+    context.parameters = parameters
```
Comment (flopezag) [inline features/environment.py:64]: This for loop is almost the same of the previous commented line.

File: features/environment.py [hunk 6]
```diff
@@ -24,16 +57,29 @@ def before_feature(context, feature):
...
+    parameters = {}
+    # parameters = dict(s.split(':', 1) for s in parameters)
+    context.parameters = parameters
+    for k,v in (s.split(':', 1) for s in p):
+        parameters[k.strip()] = v.strip()
+
+    if 'docker-compose' in parameters:
+        r = get(parameters['docker-compose'], allow_redirects=True)
+        open('docker-compose.yml', 'wb').write(r.content)
+
```
Comment (flopezag) [inline features/environment.py:71]: else:
    error!!! y fin, el docker-compose tiene que estar entre los parámetros sino se termina la fiesta...

File: features/environment.py [hunk 7]
```diff
@@ -24,16 +57,29 @@ def before_feature(context, feature):
...
+    if 'environment' in parameters:
+        r = get(parameters['environment'], allow_redirects=True)
+        open('.env', 'wb').write(r.content)
```
Comment (flopezag) [inline features/environment.py:74]: else:
    error!!! el fichero de configuración .env está marcado en los totorales para poder tener la configuración de los docker-compose, por tanto debería ser obligatorio.

File: features/environment.py [hunk 8]
```diff
@@ -24,16 +57,29 @@ def before_feature(context, feature):
...
 
-    r = get(parameters['docker-compose'], allow_redirects=True)
-    open('docker-compose.yml', 'wb').write(r.content)
+    if 'docker-compose' in parameters:
```
Comment (flopezag) [inline features/environment.py:76]: Si hemos llegado aquí sin terminar la aplicación en el caso anterior no se necesitaría comparar otra vez ...

**Review by flopezag — CHANGES_REQUESTED** [review]

File: features/environment.py (hunk 4 above)
Comment (jicarretero) [reply features/environment.py:60]: A featrue.description is a string of the form "xxxx: bla bla bla bla", I need to know if the string starts with any of the strings in "INTERESTING_FEATURES_STRINGS", so a function was added. "s" will never be in the INTERESTING_FEATURES_STRINGS.

File: .gitignore (hunk 1 above)
Comment (flopezag) [reply .gitignore:120]: NTC

File: features/environment.py (hunk 4 above)
Comment (flopezag) [reply features/environment.py:60]: NTC

**Comment by flopezag** [comment]
LGTM

**Review by flopezag — APPROVED** [review]

**Pull request merged**